packages/json-parser-products/src/main.js

```diff
@@ -0,0 +1,307 @@
+      ...logger,
+    }
+    this.categoriesCache = {}
+    this.accessToken = accessToken
+
+    const mappingParams = {
+      fillAllRows: this.parserConfig.fillAllRows,
+      categoryBy: this.parserConfig.categoryBy,
+      lang: this.parserConfig.lang,
+      multiValueDelimiter: this.parserConfig.multiValueDelimiter,
+    }
+    this._productMapping = new ProductMapping(mappingParams)
+  }
+
+  run (input: stream$Readable, output: stream$Writable) {
+    const productStream = this.parse(input, output)
+    const headers = this.parserConfig.headers
+
+    if (headers)
+      writeToSingleCsvFile(productStream, output, this.logger, headers)
+    else
+      writeToZipFile(productStream, output, this.logger)
+  }
+
+  parse (input: stream$Readable, output: stream$Writable) {
+    this.logger.info('Starting conversion')
+    let productCount = 0
+    input.setEncoding('utf8') // TODO: Move to CLI
+
+    return highland(input)
+      // parse chunk and split into JSON object strings
+      .splitBy('\n')
+      // convert the JSON object strings to JS objects
+      .map(JSON.parse)
+      .flatMap((product: ProductProjection) => (
+        highland(this._resolveReferences(product))))
+      .doto(() => {
+        productCount += 1
+        this.logger.debug(`Resolved references of ${productCount} products`)
+      })
+      // prepare the product objects for csv format
+      .map((product: ResolvedProdProj) => this._productMapping.run(product))
+      .flatten()
+      .doto(() => {
+        this.logger.debug(`Done with conversion of ${productCount} products`)
+      })
+      .stopOnError((err: Error) => {
+        this.logger.error(err)
+        output.emit('error', err)
+      })
+  }
+
+  _resolveReferences (
+    product: ProductProjection,
+    ): ResolvedProdProj {
+    // ReferenceTypes that need to be resolved:
+    // **ProductType
+    // **Categories [array]
+    // **TaxCategory
+    // **State
+    // **CategoryOrderHints
+
+    return Promise.all([
+      this._resolveProductType(product.productType),
+      this._resolveTaxCategory(product.taxCategory),
+      this._resolveState(product.state),
+      this._resolveCategories(product.categories),
+      this._resolveCategoryOrderHints(product.categoryOrderHints),
+    ])
+    .then(([
+      productType,
+      taxCategory,
+      state,
+      categories,
+      categoryOrderHints,
+    ]: Array<Object>): ResolvedProdProj => (
+      {
+        ...product,
+        ...productType,
+        ...taxCategory,
+        ...state,
+        ...categories,
+        ...categoryOrderHints,
+      }
+    ))
+  }
+
+  _resolveProductType (productTypeReference: TypeReference): Object {
+    if (!productTypeReference)
+      return {}
+
+    const productTypeService = this._createService('productTypes')
+    const uri = productTypeService.byId(productTypeReference.id).build()
+    return this.fetchReferences(uri)
+      .then(({ body }: SuccessResult): { productType: ProductType } => (
```

Comment:
As it's defined [here](https://github.com/commercetools/nodejs/blob/master/types/sdk.js#L35) I thought it wise to go with it. Each resolution will have different types and this was the most suited to cover them all. And since we have just one method handling all fetched data, we would run into flow issues if we are more specific